tchannel/container/heap.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+
+
+class HeapOperation(object):
+    def cmp(self, i, j):
+        """Compare the items in position i and j of the heap.
+
+        :param i: the first item's position of the heap list
+        :param j: the second item's position of the heap list
+        :return true if items[i] < items[j], otherwise false.
+        """
+        raise NotImplementedError()
+
+    def push(self, x):
+        """Push a new item into heap"""
+        raise NotImplementedError()
+
+    def pop(self):
+        """Pop an item from the heap"""
+        raise NotImplementedError()
+
+    def swap(self, i, j):
+        """swap items between position i and j of the heap"""
+        raise NotImplementedError()
+
+
+def init(h):
+    # heapify
+    n = len(h)
+    for i in range(n/2 - 1, -1):
+        down(h, i, n)
+
+
+def push(h, x):
+    h.push(x)
+    up(h, len(h)-1)
```

Comment:
No, what I'm trying to avoid is magic methods like `__len__`, etc. on things that may not have any definition of that. A peer cannot be less than another peer. Maybe `__len__` on a `PeerHeap` is okay because you want to determine the number of objects inside the heap.

But I would rather not have the implementation of the heap algorithm itself depend on magic methods. `HeapOperation` is the algorithm's interface to the underlying storage. It should not rely on anything implemented on the `HeapOperation` besides what was specified in the interface above. Make a `size`/`length` method a part of the `HeapOperation` interface and use that (`h.size()`/`h.length()`) instead of `len(h)`